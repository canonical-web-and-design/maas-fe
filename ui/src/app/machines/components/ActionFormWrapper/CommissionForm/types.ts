import type { ScriptInput } from "app/store/machine/types/actions";
import type { Script } from "app/store/script/types";

export type CommissionFormValues = {
  enableSSH: boolean;
  skipBMCConfig: boolean;
  skipNetworking: boolean;
  skipStorage: boolean;
  updateFirmware: boolean;
  configureHBA: boolean;
  commissioningScripts: Script[];
  testingScripts: Script[];
  scriptInputs: ScriptInput;
};

export type FormattedScript = Script & {
  displayName: string;
};
