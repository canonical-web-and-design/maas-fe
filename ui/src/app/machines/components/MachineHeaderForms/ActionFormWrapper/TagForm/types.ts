export type TagFormValues = {
  added: string[];
  removed: string[];
};
