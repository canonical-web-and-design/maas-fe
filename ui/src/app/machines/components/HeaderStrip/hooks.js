import { useEffect, useRef } from "react";

// Handle checking when a value has cycled from false to true.
const useCycled = (newState, onCycled) => {
  const previousState = useRef(newState);
  useEffect(() => {
    if (newState && !previousState.current) {
      onCycled();
    }
    if (previousState.current !== newState) {
      previousState.current = newState;
    }
  }, [newState, onCycled]);
};

export const useMachinesProcessing = (
  machinesProcessing,
  setProcessing,
  setSelectedAction,
  hasErrors = false
) => {
  const processingStarted = useRef(false);
  const selectedSavingCount = machinesProcessing.length;
  if (processingStarted.current === false && selectedSavingCount > 0) {
    processingStarted.current = true;
  }

  // If all the machines have finished processing then update the state.
  useCycled(selectedSavingCount === 0, () => {
    if (!hasErrors) {
      setProcessing(false);
      setSelectedAction(null);
    }
  });

  // If the machines are processing and there are errors then exit the
  // processing state.
  useCycled(hasErrors, () => {
    if (processingStarted) {
      setProcessing(false);
    }
  });
};
