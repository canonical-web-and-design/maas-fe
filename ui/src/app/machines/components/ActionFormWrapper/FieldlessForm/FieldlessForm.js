import { useDispatch, useSelector } from "react-redux";
import PropTypes from "prop-types";
import React from "react";

import { kebabToCamelCase } from "app/utils";
import { machine as machineActions } from "app/base/actions";
import { useMachineActionForm } from "app/machines/hooks";
import machineSelectors from "app/store/machine/selectors";
import ActionForm from "app/base/components/ActionForm";

// List of machine actions that do not require any extra parameters sent through
// the websocket apart from machine system id. All other actions will have
// their own form components.
const fieldlessActions = [
  "abort",
  "acquire",
  "delete",
  "exit-rescue-mode",
  "lock",
  "mark-broken",
  "mark-fixed",
  "off",
  "on",
  "release",
  "rescue-mode",
  "unlock",
];

export const FieldlessForm = ({ selectedAction, setSelectedAction }) => {
  const dispatch = useDispatch();
  const errors = useSelector(machineSelectors.errors);
  const { machinesToAction, processingCount } = useMachineActionForm(
    selectedAction.name
  );

  return (
    <ActionForm
      actionName={selectedAction.name}
      allowUnchanged
      cleanup={machineActions.cleanup}
      clearSelectedAction={() => setSelectedAction(null, true)}
      errors={errors}
      modelName="machine"
      onSubmit={() => {
        if (fieldlessActions.includes(selectedAction.name)) {
          const actionMethod = kebabToCamelCase(selectedAction.name);
          machinesToAction.forEach((machine) => {
            dispatch(machineActions[actionMethod](machine.system_id));
          });
        }
      }}
      processingCount={processingCount}
      selectedCount={machinesToAction.length}
      submitAppearance={
        selectedAction.name === "delete" ? "negative" : "positive"
      }
    />
  );
};

FieldlessForm.propTypes = {
  selectedAction: PropTypes.object.isRequired,
  setSelectedAction: PropTypes.func.isRequired,
};

export default FieldlessForm;
