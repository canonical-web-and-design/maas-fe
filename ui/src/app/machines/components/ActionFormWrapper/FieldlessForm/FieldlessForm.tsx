import { usePrevious } from "@canonical/react-components/dist/hooks";
import { useDispatch, useSelector } from "react-redux";
import { Redirect } from "react-router-dom";

import type { Props as ActionFormProps } from "app/base/components/ActionForm";
import ActionForm from "app/base/components/ActionForm";
import { useMachineActionForm } from "app/machines/hooks";
import type {
  SelectedAction,
  SetSelectedAction,
} from "app/machines/views/MachineDetails/types";
import { actions as machineActions } from "app/store/machine";
import machineSelectors from "app/store/machine/selectors";
import { NodeActions } from "app/store/types/node";
import { kebabToCamelCase } from "app/utils";

// List of machine actions that do not require any extra parameters sent through
// the websocket apart from machine system id. All other actions will have
// their own form components.
const fieldlessActions = [
  NodeActions.ABORT,
  NodeActions.ACQUIRE,
  NodeActions.DELETE,
  NodeActions.EXIT_RESCUE_MODE,
  NodeActions.LOCK,
  NodeActions.MARK_BROKEN,
  NodeActions.MARK_FIXED,
  NodeActions.OFF,
  NodeActions.ON,
  NodeActions.RESCUE_MODE,
  NodeActions.UNLOCK,
];

type Props = {
  actionDisabled?: ActionFormProps["actionDisabled"];
  selectedAction: SelectedAction;
  setSelectedAction: SetSelectedAction;
};

export const FieldlessForm = ({
  actionDisabled,
  selectedAction,
  setSelectedAction,
}: Props): JSX.Element => {
  const dispatch = useDispatch();
  const activeMachine = useSelector(machineSelectors.active);
  const { errors, machinesToAction, processingCount } = useMachineActionForm(
    selectedAction.name
  );
  const isDeletingMachine =
    activeMachine &&
    selectedAction.name === NodeActions.DELETE &&
    processingCount === 1;
  const previousIsDeletingMachine = usePrevious(isDeletingMachine, false);
  // Check if the machine cycled from deleting to not deleting and didn't
  // return an error.
  if (previousIsDeletingMachine && !isDeletingMachine && !errors) {
    // The machine was just deleted so redirect to the machine list.
    return <Redirect to="/machines" />;
  }

  return (
    <ActionForm
      actionDisabled={actionDisabled}
      actionName={selectedAction.name}
      allowUnchanged
      cleanup={machineActions.cleanup}
      clearSelectedAction={() => setSelectedAction(null, true)}
      errors={errors}
      modelName="machine"
      onSaveAnalytics={{
        action: "Submit",
        category: `Machine ${activeMachine ? "details" : "list"} action form`,
        label: selectedAction.name,
      }}
      onSubmit={() => {
        if (fieldlessActions.includes(selectedAction.name)) {
          const actionMethod = kebabToCamelCase(selectedAction.name);
          machinesToAction.forEach((machine) => {
            dispatch(machineActions[actionMethod](machine.system_id));
          });
        }
      }}
      processingCount={processingCount}
      selectedCount={machinesToAction.length}
      submitAppearance={
        selectedAction.name === NodeActions.DELETE ? "negative" : "positive"
      }
    />
  );
};

export default FieldlessForm;
