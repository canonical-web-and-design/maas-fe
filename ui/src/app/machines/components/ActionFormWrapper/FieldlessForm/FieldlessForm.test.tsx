import * as reactComponentHooks from "@canonical/react-components/dist/hooks";
import { mount } from "enzyme";
import { act } from "react-dom/test-utils";
import { Provider } from "react-redux";
import { MemoryRouter, Route } from "react-router-dom";
import configureStore from "redux-mock-store";

import FieldlessForm from "./FieldlessForm";

import type { RootState } from "app/store/root/types";
import { NodeActions } from "app/store/types/node";
import {
  generalState as generalStateFactory,
  machine as machineFactory,
  machineAction as machineActionFactory,
  machineActionsState as machineActionsStateFactory,
  machineEventError as machineEventErrorFactory,
  machineState as machineStateFactory,
  machineStatus as machineStatusFactory,
  rootState as rootStateFactory,
} from "testing/factories";

jest.mock("@canonical/react-components/dist/hooks", () => ({
  usePrevious: jest.fn(),
}));

const mockStore = configureStore();

describe("FieldlessForm", () => {
  let initialState: RootState;

  beforeEach(() => {
    initialState = rootStateFactory({
      machine: machineStateFactory({
        loaded: true,
        items: [
          machineFactory({
            system_id: "abc123",
          }),
        ],
        selected: [],
        statuses: {
          abc123: machineStatusFactory(),
        },
      }),
      general: generalStateFactory({
        machineActions: machineActionsStateFactory({
          data: [
            machineActionFactory({ name: NodeActions.ABORT, title: "Abort" }),
            machineActionFactory({
              name: NodeActions.ACQUIRE,
              title: "Acquire",
            }),
            machineActionFactory({ name: NodeActions.DELETE, title: "Delete" }),
            machineActionFactory({
              name: NodeActions.EXIT_RESCUE_MODE,
              title: "Exit rescue mode",
            }),
            machineActionFactory({ name: NodeActions.LOCK, title: "Lock" }),
            machineActionFactory({
              name: NodeActions.MARK_FIXED,
              title: "Mark fixed",
            }),
            machineActionFactory({ name: NodeActions.OFF, title: "Power off" }),
            machineActionFactory({ name: NodeActions.ON, title: "Power on" }),
            machineActionFactory({
              name: NodeActions.RESCUE_MODE,
              title: "Rescue mode",
            }),
            machineActionFactory({ name: NodeActions.UNLOCK, title: "Unlock" }),
          ],
        }),
      }),
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it("renders", () => {
    const state = { ...initialState };
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <FieldlessForm
            selectedAction={{ name: NodeActions.ON }}
            setSelectedAction={jest.fn()}
          />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("FieldlessForm")).toMatchSnapshot();
  });

  it("can unset the selected action", () => {
    const state = { ...initialState };
    state.machine.items = [
      machineFactory({ system_id: "a", actions: [NodeActions.ON] }),
    ];
    state.machine.selected = ["a"];
    state.machine.statuses = { a: machineStatusFactory() };
    const store = mockStore(state);
    const setSelectedAction = jest.fn();
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <FieldlessForm
            selectedAction={{ name: NodeActions.ON }}
            setSelectedAction={setSelectedAction}
          />
        </MemoryRouter>
      </Provider>
    );
    wrapper.find('[data-test="cancel-action"] button').simulate("click");

    expect(setSelectedAction).toHaveBeenCalledWith(null, true);
  });

  it("can dispatch abort action on selected machines", () => {
    const state = { ...initialState };
    state.machine.items = [
      machineFactory({ system_id: "abc123", actions: [NodeActions.ABORT] }),
    ];
    state.machine.selected = ["abc123"];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <FieldlessForm
            selectedAction={{ name: NodeActions.ABORT }}
            setSelectedAction={jest.fn()}
          />
        </MemoryRouter>
      </Provider>
    );

    act(() => wrapper.find("Formik").props().onSubmit());
    expect(
      store.getActions().filter(({ type }) => type === "machine/abort")
    ).toStrictEqual([
      {
        type: "machine/abort",
        meta: {
          model: "machine",
          method: "action",
        },
        payload: {
          params: {
            action: NodeActions.ABORT,
            extra: {},
            system_id: "abc123",
          },
        },
      },
    ]);
  });

  it("can dispatch abort action from details view", () => {
    const state = { ...initialState };
    state.machine.active = "abc123";
    state.machine.selected = [];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
        >
          <Route
            exact
            path="/machine/:id"
            component={() => (
              <FieldlessForm
                selectedAction={{ name: NodeActions.ABORT }}
                setSelectedAction={jest.fn()}
              />
            )}
          />
        </MemoryRouter>
      </Provider>
    );

    act(() => wrapper.find("Formik").props().onSubmit());
    expect(
      store.getActions().filter(({ type }) => type === "machine/abort")
    ).toStrictEqual([
      {
        type: "machine/abort",
        meta: {
          model: "machine",
          method: "action",
        },
        payload: {
          params: {
            action: NodeActions.ABORT,
            extra: {},
            system_id: "abc123",
          },
        },
      },
    ]);
  });

  it("can dispatch acquire action on selected machines", () => {
    const state = { ...initialState };
    state.machine.items = [
      machineFactory({ system_id: "abc123", actions: [NodeActions.ACQUIRE] }),
    ];
    state.machine.selected = ["abc123"];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <FieldlessForm
            selectedAction={{ name: NodeActions.ACQUIRE }}
            setSelectedAction={jest.fn()}
          />
        </MemoryRouter>
      </Provider>
    );

    act(() => wrapper.find("Formik").props().onSubmit());
    expect(
      store.getActions().filter(({ type }) => type === "machine/acquire")
    ).toStrictEqual([
      {
        type: "machine/acquire",
        meta: {
          model: "machine",
          method: "action",
        },
        payload: {
          params: {
            action: NodeActions.ACQUIRE,
            extra: {},
            system_id: "abc123",
          },
        },
      },
    ]);
  });

  it("can dispatch acquire action from details view", () => {
    const state = { ...initialState };
    state.machine.active = "abc123";
    state.machine.selected = [];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
        >
          <Route
            exact
            path="/machine/:id"
            component={() => (
              <FieldlessForm
                selectedAction={{ name: NodeActions.ACQUIRE }}
                setSelectedAction={jest.fn()}
              />
            )}
          />
        </MemoryRouter>
      </Provider>
    );

    act(() => wrapper.find("Formik").props().onSubmit());

    expect(
      store.getActions().filter(({ type }) => type === "machine/acquire")
    ).toStrictEqual([
      {
        type: "machine/acquire",
        meta: {
          model: "machine",
          method: "action",
        },
        payload: {
          params: {
            action: NodeActions.ACQUIRE,
            extra: {},
            system_id: "abc123",
          },
        },
      },
    ]);
  });

  it("can dispatch exit rescue mode action on selected machines", () => {
    const state = { ...initialState };
    state.machine.items = [
      machineFactory({
        system_id: "abc123",
        actions: [NodeActions.EXIT_RESCUE_MODE],
      }),
    ];
    state.machine.selected = ["abc123"];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <FieldlessForm
            selectedAction={{ name: NodeActions.EXIT_RESCUE_MODE }}
            setSelectedAction={jest.fn()}
          />
        </MemoryRouter>
      </Provider>
    );

    act(() => wrapper.find("Formik").props().onSubmit());
    expect(
      store.getActions().filter(({ type }) => type === "machine/exitRescueMode")
    ).toStrictEqual([
      {
        type: "machine/exitRescueMode",
        meta: {
          model: "machine",
          method: "action",
        },
        payload: {
          params: {
            action: NodeActions.EXIT_RESCUE_MODE,
            extra: {},
            system_id: "abc123",
          },
        },
      },
    ]);
  });

  it("can dispatch exit rescue mode action from details view", () => {
    const state = { ...initialState };
    state.machine.active = "abc123";
    state.machine.selected = [];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
        >
          <Route
            exact
            path="/machine/:id"
            component={() => (
              <FieldlessForm
                selectedAction={{ name: NodeActions.EXIT_RESCUE_MODE }}
                setSelectedAction={jest.fn()}
              />
            )}
          />
        </MemoryRouter>
      </Provider>
    );

    act(() => wrapper.find("Formik").props().onSubmit());

    expect(
      store.getActions().filter(({ type }) => type === "machine/exitRescueMode")
    ).toStrictEqual([
      {
        type: "machine/exitRescueMode",
        meta: {
          model: "machine",
          method: "action",
        },
        payload: {
          params: {
            action: NodeActions.EXIT_RESCUE_MODE,
            extra: {},
            system_id: "abc123",
          },
        },
      },
    ]);
  });

  it("can dispatch lock action on selected machines", () => {
    const state = { ...initialState };
    state.machine.items = [
      machineFactory({ system_id: "abc123", actions: [NodeActions.LOCK] }),
    ];
    state.machine.selected = ["abc123"];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <FieldlessForm
            selectedAction={{ name: NodeActions.LOCK }}
            setSelectedAction={jest.fn()}
          />
        </MemoryRouter>
      </Provider>
    );

    act(() => wrapper.find("Formik").props().onSubmit());
    expect(
      store.getActions().filter(({ type }) => type === "machine/lock")
    ).toStrictEqual([
      {
        type: "machine/lock",
        meta: {
          model: "machine",
          method: "action",
        },
        payload: {
          params: {
            action: NodeActions.LOCK,
            extra: {},
            system_id: "abc123",
          },
        },
      },
    ]);
  });

  it("can dispatch lock action from details view", () => {
    const state = { ...initialState };
    state.machine.active = "abc123";
    state.machine.selected = [];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
        >
          <Route
            exact
            path="/machine/:id"
            component={() => (
              <FieldlessForm
                selectedAction={{ name: NodeActions.LOCK }}
                setSelectedAction={jest.fn()}
              />
            )}
          />
        </MemoryRouter>
      </Provider>
    );

    act(() => wrapper.find("Formik").props().onSubmit());

    expect(
      store.getActions().filter(({ type }) => type === "machine/lock")
    ).toStrictEqual([
      {
        type: "machine/lock",
        meta: {
          model: "machine",
          method: "action",
        },
        payload: {
          params: {
            action: NodeActions.LOCK,
            extra: {},
            system_id: "abc123",
          },
        },
      },
    ]);
  });

  it("can dispatch mark fixed action on selected machines", () => {
    const state = { ...initialState };
    state.machine.items = [
      machineFactory({
        system_id: "abc123",
        actions: [NodeActions.MARK_FIXED],
      }),
    ];
    state.machine.selected = ["abc123"];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <FieldlessForm
            selectedAction={{ name: NodeActions.MARK_FIXED }}
            setSelectedAction={jest.fn()}
          />
        </MemoryRouter>
      </Provider>
    );

    act(() => wrapper.find("Formik").props().onSubmit());
    expect(
      store.getActions().filter(({ type }) => type === "machine/markFixed")
    ).toStrictEqual([
      {
        type: "machine/markFixed",
        meta: {
          model: "machine",
          method: "action",
        },
        payload: {
          params: {
            action: NodeActions.MARK_FIXED,
            extra: {},
            system_id: "abc123",
          },
        },
      },
    ]);
  });

  it("can dispatch mark fixed action from details view", () => {
    const state = { ...initialState };
    state.machine.active = "abc123";
    state.machine.selected = [];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
        >
          <Route
            exact
            path="/machine/:id"
            component={() => (
              <FieldlessForm
                selectedAction={{ name: NodeActions.MARK_FIXED }}
                setSelectedAction={jest.fn()}
              />
            )}
          />
        </MemoryRouter>
      </Provider>
    );

    act(() => wrapper.find("Formik").props().onSubmit());

    expect(
      store.getActions().filter(({ type }) => type === "machine/markFixed")
    ).toStrictEqual([
      {
        type: "machine/markFixed",
        meta: {
          model: "machine",
          method: "action",
        },
        payload: {
          params: {
            action: NodeActions.MARK_FIXED,
            extra: {},
            system_id: "abc123",
          },
        },
      },
    ]);
  });

  it("can dispatch power off action on selected machines", () => {
    const state = { ...initialState };
    state.machine.items = [
      machineFactory({ system_id: "abc123", actions: [NodeActions.OFF] }),
    ];
    state.machine.selected = ["abc123"];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <FieldlessForm
            selectedAction={{ name: NodeActions.OFF }}
            setSelectedAction={jest.fn()}
          />
        </MemoryRouter>
      </Provider>
    );

    act(() => wrapper.find("Formik").props().onSubmit());
    expect(
      store.getActions().filter(({ type }) => type === "machine/off")
    ).toStrictEqual([
      {
        type: "machine/off",
        meta: {
          model: "machine",
          method: "action",
        },
        payload: {
          params: {
            action: NodeActions.OFF,
            extra: {},
            system_id: "abc123",
          },
        },
      },
    ]);
  });

  it("can dispatch power off action from details view", () => {
    const state = { ...initialState };
    state.machine.active = "abc123";
    state.machine.selected = [];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
        >
          <Route
            exact
            path="/machine/:id"
            component={() => (
              <FieldlessForm
                selectedAction={{ name: NodeActions.OFF }}
                setSelectedAction={jest.fn()}
              />
            )}
          />
        </MemoryRouter>
      </Provider>
    );

    act(() => wrapper.find("Formik").props().onSubmit());

    expect(
      store.getActions().filter(({ type }) => type === "machine/off")
    ).toStrictEqual([
      {
        type: "machine/off",
        meta: {
          model: "machine",
          method: "action",
        },
        payload: {
          params: {
            action: NodeActions.OFF,
            extra: {},
            system_id: "abc123",
          },
        },
      },
    ]);
  });

  it("can dispatch power on action on selected machines", () => {
    const state = { ...initialState };
    state.machine.items = [
      machineFactory({ system_id: "abc123", actions: [NodeActions.ON] }),
    ];
    state.machine.selected = ["abc123"];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <FieldlessForm
            selectedAction={{ name: NodeActions.ON }}
            setSelectedAction={jest.fn()}
          />
        </MemoryRouter>
      </Provider>
    );

    act(() => wrapper.find("Formik").props().onSubmit());
    expect(
      store.getActions().filter(({ type }) => type === "machine/on")
    ).toStrictEqual([
      {
        type: "machine/on",
        meta: {
          model: "machine",
          method: "action",
        },
        payload: {
          params: {
            action: NodeActions.ON,
            extra: {},
            system_id: "abc123",
          },
        },
      },
    ]);
  });

  it("can dispatch power on action from details view", () => {
    const state = { ...initialState };
    state.machine.active = "abc123";
    state.machine.selected = [];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
        >
          <Route
            exact
            path="/machine/:id"
            component={() => (
              <FieldlessForm
                selectedAction={{ name: NodeActions.ON }}
                setSelectedAction={jest.fn()}
              />
            )}
          />
        </MemoryRouter>
      </Provider>
    );

    act(() => wrapper.find("Formik").props().onSubmit());

    expect(
      store.getActions().filter(({ type }) => type === "machine/on")
    ).toStrictEqual([
      {
        type: "machine/on",
        meta: {
          model: "machine",
          method: "action",
        },
        payload: {
          params: {
            action: NodeActions.ON,
            extra: {},
            system_id: "abc123",
          },
        },
      },
    ]);
  });

  it("can dispatch unlock action on selected machines", () => {
    const state = { ...initialState };
    state.machine.items = [
      machineFactory({ system_id: "abc123", actions: [NodeActions.UNLOCK] }),
    ];
    state.machine.selected = ["abc123"];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <FieldlessForm
            selectedAction={{ name: NodeActions.UNLOCK }}
            setSelectedAction={jest.fn()}
          />
        </MemoryRouter>
      </Provider>
    );

    act(() => wrapper.find("Formik").props().onSubmit());
    expect(
      store.getActions().filter(({ type }) => type === "machine/unlock")
    ).toStrictEqual([
      {
        type: "machine/unlock",
        meta: {
          model: "machine",
          method: "action",
        },
        payload: {
          params: {
            action: NodeActions.UNLOCK,
            extra: {},
            system_id: "abc123",
          },
        },
      },
    ]);
  });

  it("can dispatch unlock action from details view", () => {
    const state = { ...initialState };
    state.machine.active = "abc123";
    state.machine.selected = [];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
        >
          <Route
            exact
            path="/machine/:id"
            component={() => (
              <FieldlessForm
                selectedAction={{ name: NodeActions.UNLOCK }}
                setSelectedAction={jest.fn()}
              />
            )}
          />
        </MemoryRouter>
      </Provider>
    );

    act(() => wrapper.find("Formik").props().onSubmit());

    expect(
      store.getActions().filter(({ type }) => type === "machine/unlock")
    ).toStrictEqual([
      {
        type: "machine/unlock",
        meta: {
          model: "machine",
          method: "action",
        },
        payload: {
          params: {
            action: NodeActions.UNLOCK,
            extra: {},
            system_id: "abc123",
          },
        },
      },
    ]);
  });

  describe("delete", () => {
    it("displays a negative submit button if selected action is delete", () => {
      const state = { ...initialState };
      state.machine.items = [
        machineFactory({ system_id: "abc123", actions: [NodeActions.DELETE] }),
      ];
      state.machine.selected = ["abc123"];
      const store = mockStore(state);
      const setSelectedAction = jest.fn();
      const wrapper = mount(
        <Provider store={store}>
          <MemoryRouter
            initialEntries={[{ pathname: "/machines", key: "testKey" }]}
          >
            <FieldlessForm
              selectedAction={{ name: NodeActions.DELETE }}
              setSelectedAction={setSelectedAction}
            />
          </MemoryRouter>
        </Provider>
      );
      expect(wrapper.find("ActionButton").prop("appearance")).toBe("negative");
    });

    it("can dispatch delete action on selected machines", () => {
      const state = { ...initialState };
      state.machine.items = [
        machineFactory({ system_id: "abc123", actions: [NodeActions.DELETE] }),
      ];
      state.machine.selected = ["abc123"];
      const store = mockStore(state);
      const wrapper = mount(
        <Provider store={store}>
          <MemoryRouter
            initialEntries={[{ pathname: "/machines", key: "testKey" }]}
          >
            <FieldlessForm
              selectedAction={{ name: NodeActions.DELETE }}
              setSelectedAction={jest.fn()}
            />
          </MemoryRouter>
        </Provider>
      );

      act(() => wrapper.find("Formik").props().onSubmit());
      expect(
        store.getActions().filter(({ type }) => type === "machine/delete")
      ).toStrictEqual([
        {
          type: "machine/delete",
          meta: {
            model: "machine",
            method: "action",
          },
          payload: {
            params: {
              action: NodeActions.DELETE,
              extra: {},
              system_id: "abc123",
            },
          },
        },
      ]);
    });

    it("can dispatch delete action from details view", () => {
      const state = { ...initialState };
      state.machine.active = "abc123";
      state.machine.selected = [];
      const store = mockStore(state);
      const wrapper = mount(
        <Provider store={store}>
          <MemoryRouter
            initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
          >
            <Route
              exact
              path="/machine/:id"
              component={() => (
                <FieldlessForm
                  selectedAction={{ name: NodeActions.DELETE }}
                  setSelectedAction={jest.fn()}
                />
              )}
            />
          </MemoryRouter>
        </Provider>
      );

      act(() => wrapper.find("Formik").props().onSubmit());

      expect(
        store.getActions().filter(({ type }) => type === "machine/delete")
      ).toStrictEqual([
        {
          type: "machine/delete",
          meta: {
            model: "machine",
            method: "action",
          },
          payload: {
            params: {
              action: NodeActions.DELETE,
              extra: {},
              system_id: "abc123",
            },
          },
        },
      ]);
    });

    it("redirects when a machine is deleted", () => {
      const state = { ...initialState };
      state.machine.active = "abc123";
      state.machine.selected = [];
      state.machine.statuses.abc123.deleting = false;
      jest
        .spyOn(reactComponentHooks, "usePrevious")
        .mockImplementation(() => true);
      const store = mockStore(state);
      const wrapper = mount(
        <Provider store={store}>
          <MemoryRouter
            initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
          >
            <FieldlessForm
              selectedAction={{ name: NodeActions.DELETE }}
              setSelectedAction={jest.fn()}
            />
          </MemoryRouter>
        </Provider>
      );
      expect(wrapper.find("Redirect").exists()).toBe(true);
    });

    it("does not redirect if there are errors", () => {
      const state = { ...initialState };
      state.machine.active = "abc123";
      state.machine.selected = [];
      state.machine.statuses.abc123.deleting = false;
      state.machine.eventErrors = [
        machineEventErrorFactory({
          id: "abc123",
          event: "delete",
          error: "uh oh",
        }),
      ];
      jest
        .spyOn(reactComponentHooks, "usePrevious")
        .mockImplementation(() => true);
      const store = mockStore(state);
      const wrapper = mount(
        <Provider store={store}>
          <MemoryRouter
            initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
          >
            <FieldlessForm
              selectedAction={{ name: NodeActions.DELETE }}
              setSelectedAction={jest.fn()}
            />
          </MemoryRouter>
        </Provider>
      );
      expect(wrapper.find("Redirect").exists()).toBe(false);
    });
  });
});
