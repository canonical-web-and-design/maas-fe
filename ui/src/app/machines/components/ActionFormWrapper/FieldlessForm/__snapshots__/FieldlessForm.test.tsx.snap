// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldlessForm renders 1`] = `
<FieldlessForm
  selectedAction={
    Object {
      "name": "on",
    }
  }
  setSelectedAction={[MockFunction]}
>
  <ActionForm
    actionName="on"
    allowUnchanged={true}
    cleanup={[Function]}
    clearSelectedAction={[Function]}
    modelName="machine"
    onSaveAnalytics={
      Object {
        "action": "Submit",
        "category": "Machine list action form",
        "label": "on",
      }
    }
    onSubmit={[Function]}
    processingCount={0}
    selectedCount={0}
    submitAppearance="positive"
  >
    <FormikForm
      allowAllEmpty={false}
      allowUnchanged={true}
      buttons={[Function]}
      buttonsBordered={false}
      cleanup={[Function]}
      initialValues={Object {}}
      onCancel={[Function]}
      onSaveAnalytics={
        Object {
          "action": "Submit",
          "category": "Machine list action form",
          "label": "on",
        }
      }
      onSubmit={[Function]}
      saved={false}
      saving={false}
      savingLabel="Powering on machine..."
      submitAppearance="positive"
      submitLabel="Power on machine"
    >
      <Formik
        initialValues={Object {}}
        onSubmit={[Function]}
      >
        <FormikFormContent
          allowAllEmpty={false}
          allowUnchanged={true}
          buttons={[Function]}
          buttonsBordered={false}
          editable={true}
          initialValues={Object {}}
          onCancel={[Function]}
          saved={false}
          saving={false}
          savingLabel="Powering on machine..."
          submitAppearance="positive"
          submitLabel="Power on machine"
        >
          <Form
            onSubmit={[Function]}
          >
            <form
              className=""
              onSubmit={[Function]}
            >
              <FormCardButtons
                bordered={false}
                cancelDisabled={false}
                loading={false}
                loadingLabel="Powering on machine..."
                onCancel={[Function]}
                submitAppearance="positive"
                submitDisabled={false}
                submitLabel="Power on machine"
                success={false}
              >
                <div
                  className="form-card__buttons"
                >
                  <Button
                    appearance="base"
                    className=""
                    data-test="cancel-action"
                    onClick={[Function]}
                    type="button"
                  >
                    <button
                      className="p-button--base"
                      data-test="cancel-action"
                      onClick={[Function]}
                      type="button"
                    >
                      Cancel
                    </button>
                  </Button>
                  <ActionButton
                    appearance="positive"
                    className=""
                    disabled={false}
                    loading={false}
                    success={false}
                    type="submit"
                  >
                    <button
                      className="p-button--positive"
                      disabled={false}
                      type="submit"
                    >
                      Power on machine
                    </button>
                  </ActionButton>
                </div>
              </FormCardButtons>
            </form>
          </Form>
        </FormikFormContent>
      </Formik>
    </FormikForm>
  </ActionForm>
</FieldlessForm>
`;
